File: OneVsRestClassifier.py
from creme import optim
from creme import linear_model
from creme import multiclass
import pickle
import json

#read configurations
config = json.loads(input())
init = 0

savePath = config['savePath']
target = config['target']
l2 = config['l2']
opt = config['optimizer']
lr = config['lr']
AdaBound_lr = config['AdaBound_lr']
rho = config['rho']
beta_1 = config['beta_1']
beta_2 = config['beta_2']
eps =  config['eps']
gamma = config['gamma']
final_lr = config['final_lr']
alpha = config['alpha']
FTRL_l1 =  config['FTRL_l1']
FTRL_l2 = config['FTRL_l2']


if (opt == "AdaBound"):
	optimizer = optim.AdaBound(lr, beta_1, beta_2, eps, gamma, final_lr)
elif (opt == "AdaDelta"):
	optimizer = optim.AdaDelta(rho, eps)
elif (opt == "AdaGrad"):
	optimizer = optim.AdaGrad(lr, eps)
elif (opt == "Adam"):
	optimizer = optim.Adam(lr, beta_1, beta_2, eps)
elif (opt == "FTRLProximal"):
	optimizer = optim.FTRLProximal(alpha, beta, l1, l2)
elif (opt == "Momentum"):
	optimizer = optim.Momentum(lr, rho)
elif (opt == "RMSProp"):
	optimizer = optim.RMSProp(lr, rho, eps)
elif (opt == "VanillaSGD"):
	optimizer = optim.VanillaSGD(lr)
elif (opt == "NesterovMomentum"):
	optimizer = optim.NesterovMomentum(lr, rho)
else:
	optimizer = None

log_reg = linear_model.LogisticRegression(optimizer, l2= l2)
OVRClassifier = multiclass.OneVsRestClassifier(log_reg)

output = {}

while True:

	#wait request
	data = input()

	Xi = json.loads(data)
	y = float(Xi.pop(target))

	output["Predict"] = OVRClassifier.predict_one(Xi)
	output["Truth"] = y

	model = OVRClassifier.fit_one(Xi, y)
	pickle.dump(model, open(savePath, 'wb'))

	print(json.dumps(output))

File: OneVsRestClassifier.js
	 
module.exports = function(RED){
	function OneVsRestClassifierNode(config){
		const path = require('path')
		const utils = require('./utils/utils')

		var node = this;

		//set configurations
		node.file = __dirname + '/OneVsRestClassifier.py'
		node.config = {
			savePath: config.savePath || "OneVsRestClassifier.sav",
			target: config.target || 'y',
			l2: Number(config.l2) || 0,
			optimizer: config.optimizer,
			lr: Number(config.lr) || 0.1,
			AdaBound_lr: Number(config.AdaBound_lr) || 0.001,
			rho: Number(config.rho) || 0.9,
			beta_1: Number(config.beta_1) || 0.9,
			beta_2: Number(config.beta_2) || 0.999,
			eps: Number(config.eps) || 1e-08,
			gamma: Number(config.gamma) || 0.001,
			final_lr: Number(config.final_lr) || 0.1,
			alpha: Number(config.alpha) || 0.05,
			beta: Number(config.beta) || 1.0,
			FTRL_l1: Number(config.FTRL_l1) || 0.0,
			FTRL_l2: Number(config.FTRL_l2) || 1.0
		}
		utils.run(RED, node, config)
	}
	RED.nodes.registerType("OneVsRestClassifier", OneVsRestClassifierNode)
}
